src/common/util/util.ts

```diff
@@ -219,6 +219,50 @@ export function mapLazy<T, R>(xs: Iterable<T>, f: (x: T) => R): Iterable<R> {
   };
 }
 
+const Orders = {
+  forward: true,
+  backward: true,
+  random: true,
+};
+export type Order = keyof typeof Orders;
+export const kOrderKeys = keysOf(Orders);
+
+/**
+ * Returns a random integer in the range [min, max)
+ * min and max must be >= 0
+ */
+function randInt(min: number, max: number): number {
+  return (min + Math.random() * (max - min)) | 0;
+}
+
+/**
+ * Shuffles an array's elements in place.
+ */
+export function shuffleInPlace<R>(arr: R[]): R[] {
+  const len = arr.length;
+  for (let i = 0; i < len - 1; ++i) {
+    const other = randInt(i, len);
+    [arr[i], arr[other]] = [arr[other], arr[i]];
+  }
+  return arr;
+}
+
+/**
+ * Reorders an array, in place, based on the Order
+ */
+export function reorder<R>(order: Order, arr: R[]): R[] {
```

Comment:
nit: Since this modifies the array in place I think it would be clearer if it didn't return it back.

Reply:
switched to returning a new array. It's more convenient to use